File: [0545][Medium][Boundary of Binary Tree]/Boundary of Binary Tree.java
// leetcode time     cost : 2 ms
// leetcode memory   cost : 40 MB 
// Time  Complexity: O(n)
// Space Complexity: O(n)

/**
 * Definition for a binary tree node.
 */
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

// # solution 1, 3 times loop for left boundary, leaves, right boundary,
public class Solution {

    public boolean isLeaf(TreeNode t) {
        return t.left == null && t.right == null;
    }

    public void addLeaves(List<Integer> res, TreeNode root) {
        if (isLeaf(root)) {
            res.add(root.val);
        } else {
            if (root.left != null) {
                addLeaves(res, root.left);
            }
            if (root.right != null) {
                addLeaves(res, root.right);
            }
        }
    }

    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        ArrayList<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        if (!isLeaf(root)) {
            res.add(root.val);
        }
        TreeNode tmp = root.left;
        while (tmp != null) {
            if (!isLeaf(tmp)) {
                res.add(tmp.val);
            }
            if (tmp.left != null) {
                tmp = tmp.left;
            } else {
                tmp = tmp.right;
            }

        }
        addLeaves(res, root);
        Stack<Integer> s = new Stack<>();
        tmp = root.right;
        while (tmp != null) {
            if (!isLeaf(tmp)) {
                s.push(tmp.val);
            }
            if (tmp.right != null) {
                tmp = tmp.right;
            } else {
                tmp = tmp.left;
            }
        }
        while (!s.empty()) {
            res.add(s.pop());
        }
        return res;
    }
}

File: [0545][Medium][Boundary of Binary Tree]/Boundary of Binary Tree.py
# leetcode time     cost : 48 ms
# leetcode memory   cost : 16.1 MB
# Time  Complexity: O(n)
# Space Complexity: O(n)

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# solution 1, 3 times loop for left boundary, leaves, right boundary,


class Solution:
    def boundaryOfBinaryTree(self, root: TreeNode):
        left_b, right_b, leaves = [], [], []

        def get_left_boundary(node):
            if not node:
                return
            left_b.append(node.val)
            if node.left:
                get_left_boundary(node.left)
            else:
                get_left_boundary(node.right)

        def get_right_boundary(node):
            if not node:
                return
            right_b.append(node.val)
            if node.right:
                get_right_boundary(node.right)
            else:
                get_right_boundary(node.left)

        def get_leaves(node):
            if not node:
                return None
            if not node.left and not node.right:
                leaves.append(node.val)
                return
            get_leaves(node.left)
            get_leaves(node.right)

        if root == None:
            return []
        if not root.left and not root.right:
            return [root.val]
        if root.left:
            get_left_boundary(root.left)
            # discard the duplicate node already scanned in the leaves
            if len(left_b) > 0:
                left_b.pop()
        if root.right:
            get_right_boundary(root.right)
            # discard the duplicate node already scanned in the leaves
            if len(right_b) > 0:
                right_b.pop()
        get_leaves(root)
        return [root.val] + left_b + leaves + right_b[::-1]
